### PR TITLE
Bringing templates to feature parity in tape mode

### PR DESCRIPTION
**Pull request by mariaschuld**
Work in progress: Make the template integration tests pass in tape mode.

### CODE REVIEW

File: pennylane/tape/interfaces/functions.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
Comment (mariaschuld) [inline pennylane/tape/interfaces/functions.py:1]: For now this file is just supposed to collect all wrapper functions that seem absolutely necessary.

File: pennylane/templates/broadcast.py [hunk 2]
```diff
@@ -487,24 +489,11 @@ def circuit(pars):
                 circuit(pars)
     """
 
-    OPTIONS = ["single", "double", "double_odd", "chain", "ring", "pyramid", "all_to_all", "custom"]
```
Comment (mariaschuld) [inline pennylane/templates/broadcast.py:490]: Sorry...cleaned this file up a bit....

File: pennylane/templates/broadcast.py [hunk 3]
```diff
@@ -538,33 +540,22 @@ def circuit(pars):
...
-        "ring": wires_ring(wires),
-        "pyramid": wires_pyramid(wires),
-        "all_to_all": wires_all_to_all(wires),
-        "custom": custom_pattern,
-    }
+            parameters = qml.tape.interfaces.functions.WrapperFunctions.expand_dims(parameters, 1)
 
-    # broadcast the unitary
-    for wires, pars in zip(wire_sequence[pattern], parameters):
-        unitary(*pars, wires=wires, **kwargs)
```
Comment (mariaschuld) [inline pennylane/templates/broadcast.py:570]: Below version is cleaner, because it does not replace no parameters with empty parameters.

**Comment by mariaschuld** [comment]
[ch1036]

**Comment by mariaschuld** [comment]
@josh146 

I hoped I could finish template feature parity, but it is much harder than expected and requires a lot of interplay with the new wrapper functions. In addition to the details below I left comments in relevant parts of the code.

Here is what I got: 

* All templates now have the entire pre-processing happening in a seperate function called `_preprocess()`. Since we will soon only have tape mode, I did not bother with adapting it to both modes; there is a `if qml.tape_mode_enabled()` fork that contains either the new or the old preprocessing code. I took the liberty to polish some error messages so they read nicer. The idea is that once tape is core, the old branch can just be deleted.

* These preprocessing functions sometimes return inputs after preprocessing them (in which case they are unboxed), but most often they just inspect them in the interface agnostic manner.

* In some templates (the qchem ones) a little bit of processing happens later, but as far as I can see it is only using simple operations that all interfaces support.

* I added the concat function to the tensorbox module. But there is still a problem (which I wonder why it doesn't appear at the very similar 'stack' function): A tf test fails because concatenating tensors created from mixed types will fail due to the tensors having mixed `dtype`. I am sure this is fixable, but it entails a bit of a decision of what the expected behaviour is, so I left a note.

Here is what is missing, and which causes an awful lot of tests to fail:

* `AmplitudeEmbedding` and `MottonenStatePrep` are not yet entirely refactored. The reason is that I am missing a lot of functions. Mostly to check the norm (abs, sum, isclose) and to extract angles (angle). These templates require tensor boxes/unboxing to be used deep in the code, so it is an excellent test bed. I think that if the refactor is complete, they should be diffable. Note: these embeddings make also VQE tests fail.

* I was at a loss how to treat inputs that are scalars. Of course, when the user makes it diffable it should be something like a `np.array(1.)`, and the tensorbox works. But another user that just feeds `1.` makes everything crash. Could we add that conversion to the tensorbox? This appears in some qchem templates.

I hope that was it. This is still a big amount of work, but a good exercise to test the tensorbox module!

File: pennylane/templates/embeddings/amplitude.py [hunk 4]
```diff
@@ -15,27 +15,128 @@
...
+        if pad_with is not None and n_features < 2 ** len(wires):
+            padding = [pad_with] * (2 ** len(wires) - n_features)
+            features = qml.tensorbox.concatenate([features, padding], axis=0)
+
+        #TODO: add those methods to tensorbox
+
+        # normalize
+        norm = np.sum(np.abs(features) ** 2)
+
+        if not np.isclose(norm, 1.0, atol=TOLERANCE):
```
Comment (mariaschuld) [inline pennylane/templates/embeddings/amplitude.py:62]: This needs to be refactored to agnostic processing.

File: pennylane/templates/embeddings/amplitude.py [hunk 5]
```diff
@@ -83,28 +180,27 @@ def circuit(f=None):
...
 
-        >>> dev._state
+        >>> dev.state
         [0.5+0.j 0.5+0.j 0.5+0.j 0.5+0.j]
 
-        **Passing features as positional arguments to a quantum node**
+        **Differentiating with respect to the features**
 
-        The ``features`` argument of ``AmplitudeEmbedding`` can in principle also be passed to the quantum node
-        as a positional argument:
```
Comment (mariaschuld) [inline pennylane/templates/embeddings/amplitude.py:94]: Realised that the docstring is outdated.

File: pennylane/templates/subroutines/double_excitation_unitary.py [hunk 6]
```diff
@@ -26,6 +27,36 @@
...
+        )
+    if len(wires2) < 2:
+        raise ValueError(
+            "expected at least two wires representing the unoccupied orbitals; "
+            "got {}".format(len(wires2))
+        )
+
+    if qml.tape_mode_active():
+
+        weight = qml.tensorbox.TensorBox(np.array(weight))  # Todo: make TensorBox accept scalars
```
Comment (mariaschuld) [inline pennylane/templates/subroutines/double_excitation_unitary.py:46]: This is just a hack which causes test errors. Need to "tensorbox" scalars here.

File: pennylane/templates/subroutines/single_excitation_unitary.py [hunk 7]
```diff
@@ -26,6 +27,27 @@
...
 
+def _preprocess(weight, wires):
+    """Validate and pre-process inputs."""
+
+    if len(wires) < 2:
+        raise ValueError("expected at least two wires; got {}".format(len(wires)))
+
+    if qml.tape_mode_active():
+
+        weight = qml.tensorbox.TensorBox(np.array(weight)) # Todo: make TensorBox accept scalars
```
Comment (mariaschuld) [inline pennylane/templates/subroutines/single_excitation_unitary.py:38]: Same here.

File: pennylane/tensorbox/tf_box.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        return TensorFlowBox(tf.cast(self.data, dtype))
+
+    @staticmethod
+    def concatenate(values, axis=0):
+
+        if axis is None:
+            # flatten and then concatenate zero'th dimension
+            # to reproduce numpy's behaviour
+            tensors = [tf.reshape(TensorFlowBox.astensor(t), shape=[-1]) for t in TensorFlowBox.unbox_list(values)]
+            # TODO: error is raised when dtypes are not the same
```
Comment (mariaschuld) [inline pennylane/tensorbox/tf_box.py:54]: This is the dtype problem: `tf.concat` complains if the tensors do not have the same type. But which one is the correct one here?

File: tests/templates/test_layers.py [hunk 9]
```diff
@@ -237,55 +240,6 @@ def test_strong_ent_layers_uses_correct_number_of_imprimitives(self, n_layers, n
...
-
-        def circuit(weights):
-            StronglyEntanglingLayers(weights=weights, wires=range(n_wires), ranges=ranges)
-            return qml.expval(qml.PauliZ(0))
-
-        qnode = qml.QNode(circuit, dev)
-
-        with pytest.raises(ValueError, match="'ranges' must be of shape"):
-            qnode(weights)
-
```
Comment (mariaschuld) [inline tests/templates/test_layers.py:288]: I tried not to delete many tests, but our new design principle of not testing each and every input will make some tests obsolete. When I foresaw this to happen, and the test would need a rewrite, I just delete them.

File: pennylane/templates/embeddings/amplitude.py [hunk 10]
```diff
@@ -15,27 +15,128 @@
...
 from pennylane.templates.decorator import template
 from pennylane.ops import QubitStateVector
-from pennylane.templates.utils import (
-    check_shape,
-    check_type,
-    get_shape,
-)
 from pennylane.variable import Variable
 from pennylane.wires import Wires
+from pennylane.templates.utils import check_shape, get_shape, check_type
```
Comment (mariaschuld) [inline pennylane/templates/embeddings/amplitude.py:26]: Can't wait until we can delete all those!

File: pennylane/templates/embeddings/amplitude.py (hunk 4 above)
Comment (mariaschuld) [reply pennylane/templates/embeddings/amplitude.py:62]: Since we just want to extract the scalar norm, one could also just extract a numpy representation of the input here (for feature parity at least, I think this stops us from being diffable). But I had trouble doing that, `features.numpy()` did not work in all cases for some reason.

**Comment by josh146** [comment]
With the latest commit, all tests should now pass! As a result, all templates, including the notorious Mottonen state prep, are fully differentiable:

```python
import pennylane as qml
from pennylane import numpy as np

n_wires = 3
dev = qml.device("default.qubit", wires=n_wires)

qml.enable_tape()

@qml.qnode(dev)
def circuit(weights, init_state):
    qml.templates.MottonenStatePreparation(init_state, wires=range(n_wires))
    qml.templates.StronglyEntanglingLayers(weights, wires=range(n_wires))
    return qml.expval(qml.PauliX(0) @ qml.PauliZ(1)), qml.expval(qml.PauliZ(1))

init_state = np.array([1, 1, 1, 1, 1, 1, 1, 1], requires_grad=True)
init_state = init_state / np.linalg.norm(init_state)
weights = qml.init.strong_ent_layers_normal(n_wires=n_wires, n_layers=3)

def cost(weights, init_state):
    return np.sum(circuit(weights, init_state))

res = cost(weights, init_state)

print("Cost:", res)

grad_fn = qml.grad(cost)
print("Gradient:", grad_fn(weights, init_state))
```

**Comment by josh146** [comment]
Edit: there are two JAX tests failing, because the `TensorBox` class does not yet have JAX support 😢 This is probably a separate PR

File: pennylane/templates/embeddings/amplitude.py [hunk 11]
```diff
@@ -15,27 +15,131 @@
...
+        if pad_with is not None and n_features < 2 ** len(wires):
+            padding = [pad_with] * (2 ** len(wires) - n_features)
+            features = qml.math.concatenate([features, padding], axis=0)
+
+        # normalize
+        norm = qml.math.sum(qml.math.abs(features) ** 2)
+
+        if not qml.math.allclose(norm, 1.0, atol=TOLERANCE):
+            if normalize or pad_with:
+                features = features / np.sqrt(norm)
```
Comment (Thenerdstation) [inline pennylane/templates/embeddings/amplitude.py:66]: As discussed, JAX abstract types will break on conditional branching based on tensor values.

File: pennylane/templates/layer.py [hunk 12]
```diff
@@ -15,9 +15,27 @@
...
+    """Validate and pre-process inputs."""
+
+    if not isinstance(depth, int):
+        raise ValueError("'depth' must be of type int, got {}".format(type(depth).__name__))
+
+    for arg in args:
+
+        if qml.tape_mode_active():
+            arg = qml.math.angle(arg)
+
```
Comment (Thenerdstation) [inline pennylane/templates/layer.py:32]: Else?

Comment (josh146) [reply pennylane/templates/layer.py:32]: Looking at the 'before', it doesn't seem to be needed 🤔 

@mariaschuld, can you remember why we need this in tape mode, but not in non-tape mode?

File: pennylane/templates/broadcast.py [hunk 13]
```diff
@@ -64,6 +66,79 @@ def wires_all_to_all(wires):
 ###################
 
 
+def _preprocess(parameters, pattern, wires):
+    """Validate and pre-process inputs."""
```
Comment (Thenerdstation) [inline pennylane/templates/broadcast.py:70]: Add function argument documentation?

**Review by Thenerdstation — APPROVED** [review]
LGTM. Only comments are the breakages in tf graph mode or in a jax.jit.

File: pennylane/templates/broadcast.py (hunk 13 above)
Comment (mariaschuld) [reply pennylane/templates/broadcast.py:70]: True, that would be good. I was sloppy here...

Comment (mariaschuld) [reply pennylane/templates/broadcast.py:70]: @josh146  you want me to add them?

Comment (mariaschuld) [reply pennylane/templates/broadcast.py:70]: We could also be more precise about what preprocessing does in each case...

File: pennylane/templates/embeddings/amplitude.py (hunk 11 above)
Comment (mariaschuld) [reply pennylane/templates/embeddings/amplitude.py:66]: We need a general strategy how to check numerical properties of inputs then, for example that inputs of BasisEmbedding are 0/1, that inputs to AmplitudeEmbedding are normalised etc...

File: pennylane/templates/broadcast.py (hunk 13 above)
Comment (josh146) [reply pennylane/templates/broadcast.py:70]: Yeah, actually, it might be a good idea. Sorry I'm so used to this PR, I didn't even notice the docstrings were missing!

File: pennylane/templates/embeddings/amplitude.py (hunk 11 above)
Comment (josh146) [reply pennylane/templates/embeddings/amplitude.py:66]: * For basis embedding, since it is not differentiable, what if we restricted it to only NumPy arrays? Would the computational graph just treat is as a constant?

* For amplitude embedding, one solution is to simply always normalize (e.g., just get rid of the if statement).

File: tests/devices/test_default_qubit_jax.py [hunk 14]
```diff
@@ -439,11 +440,13 @@ def test_non_backprop_error(self):
...
 
         def circuit(weights):
-            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
             return qml.expval(qml.PauliZ(0))
 
         qnode = qml.QNode(circuit, dev, interface="jax", diff_method="parameter-shift")
-        weights = jnp.array(qml.init.strong_ent_layers_normal(n_wires=2, n_layers=2))
```
Comment (mariaschuld) [inline tests/devices/test_default_qubit_jax.py:446]: Interesting. Why this change? Just to remove a dependency?

Comment (josh146) [reply tests/devices/test_default_qubit_jax.py:446]: Yep, this test is testing JAX differentiation, so shouldn't need to depend on extra template code. This narrows the test to ensure that it passes if JAX differentiation works as intended. This way, if the test fails, we know for sure why (as opposed to it potentially being a template issue).

However, we should add a separate test to ensure that _templates_ can be differentiated.

**Review by josh146 — APPROVED** [review]
